onnxruntime/core/optimizer/nhwc_transformer.cc

```diff
@@ -0,0 +1,366 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Copyright (c) 2020, NXP Semiconductor, Inc. All rights reserved.
+// Licensed under the MIT License.
+
+#if defined(USE_ACL) || defined(USE_ARMNN)
+
+#include <deque>
+#include "core/graph/graph_utils.h"
+#include "core/optimizer/initializer.h"
+#include "core/optimizer/nhwc_transformer.h"
+
+#include "core/providers/acl/acl_common.h"
+
+// ACL
+#include "arm_compute/runtime/Tensor.h"
+#include "arm_compute/runtime/TensorAllocator.h"
+
+// NEON
+#include "arm_compute/runtime/NEON/functions/NEPermute.h"
+
+using namespace ONNX_NAMESPACE;
+using namespace ::onnxruntime::common;
+namespace onnxruntime {
+
+enum DataLayout {
+  NchwLayout,
+  NhwcLayout,
+  InheritLayout
+};
+
+class NhwcTransformerImpl {
+ public:
+  NhwcTransformerImpl(Graph& graph, std::string provider, const logging::Logger& _logger) noexcept :
+    graph_(graph), provider_(provider), logger(_logger) {};
+  void Transform(Node& node);
+  void Finalize(bool& modified);
+
+ private:
+  NodeIndex InsertPermuteParentNode(Node& node, const Node::EdgeEnd* edge, bool bNHWC);
+  NodeIndex InsertPermuteChildNode(Node& node, bool bNHWC);
+  NodeIndex ReplaceNode(Node& node);
+  void PermuteWeights(NodeArg* src, NodeArg** dst, const std::string&);
+
+  DataLayout RequiredLayout(const Node& node);
+  bool SuportsReplacementNHWC(const Node& node);
+  bool RequiresWeightsPermutation(const Node& node);
+
+  std::map<NodeIndex, bool> nodes_layout;
+  Graph& graph_;
+  std::string provider_;
+  const logging::Logger& logger;
+};
+
+NodeIndex NhwcTransformerImpl::InsertPermuteParentNode(Node& node, const Node::EdgeEnd* edge, bool bNHWC) {
+
+  const Node* parent_node = NULL;
+  int srcIdx = 0, dstIdx = 0;
+  if (edge) {
+    parent_node = &edge->GetNode();
+    srcIdx = edge->GetSrcArgIndex();
+    dstIdx = edge->GetDstArgIndex();
+  }
+
+  LOGS(logger, VERBOSE) << "NHWC Insert: "
+      << (parent_node ? parent_node->OpType() : "*")
+      << ":" << srcIdx
+      << (bNHWC ? " [ReorderInput:0] " : " [ReorderOutput:0] ")
+      << node.OpType()
+      << ":" << dstIdx;
+
+  auto& input_defs = node.MutableInputDefs();
+  std::string new_input_def_name = graph_.GenerateNodeArgName("input");
+  auto* new_input_arg = &graph_.GetOrCreateNodeArg(new_input_def_name, input_defs[0]->TypeAsProto());
```

Comment:
nit: could the original input name be used as the base for the new name?

Reply:
We can use any base name since it will be postfixed.

Reply:
Was more for debuggability. Either will work, but slightly easier to see the relationship with the original name if it's part of the new name. 

---
In reply to: [441399513](https://github.com/microsoft/onnxruntime/pull/4237#discussion_r441399513) [](ancestors = 441399513)